packages/language-service/src/ts_utils.ts

```diff
@@ -129,3 +129,165 @@ export function updateObjectValueForKey(
     newProp,
   ]);
 }
+
+/**
+ * Create a new ArrayLiteralExpression, or accept an existing one.
+ * Ensure the array contains the provided identifier.
+ * Returns the array, either updated or newly created.
+ * If no update is needed, returns `null`.
+ */
+export function ensureArrayWithIdentifier(
+    identifier: ts.Identifier, arr?: ts.ArrayLiteralExpression): ts.ArrayLiteralExpression|null {
+  if (!arr) {
+    return ts.factory.createArrayLiteralExpression([identifier]);
+  }
+  if (arr.elements.find(v => ts.isIdentifier(v) && v.text === identifier.text)) {
+    return null;
+  }
+  return ts.factory.updateArrayLiteralExpression(arr, [...arr.elements, identifier]);
+}
+
+/**
+ * Determine whether a given module specifier points to the exact path provided.
+ */
+export function moduleSpecifierHasPath(moduleSpecifier: ts.Expression, path: string): boolean {
+  return ts.isStringLiteral(moduleSpecifier) && moduleSpecifier.text === path;
+}
+
+/**
+ * Determine whether this an import of the given `propertyName` from a particular module
+ * specifier already exists. If so, return the local name for that import, which might be an
+ * alias.
+ */
+export function hasImport(
+    importDeclarations: ts.ImportDeclaration[], propName: string, moduleSpecifier: string): string|
+    null {
+  return importDeclarations
+             .filter(
+                 declaration =>
+                     moduleSpecifierHasPath(declaration.moduleSpecifier, moduleSpecifier))
+             .map(decl => importHas(decl, propName))
+             .find(p => p !== null) ??
+      null;
+}
+
+/**
+ * Determine whether this import declaration already contains an import of the given
+ * `propertyName`, and if so, the name it can be referred to with in the local scope.
+ */
+function importHas(importDecl: ts.ImportDeclaration, propName: string): string|null {
+  const bindings = importDecl.importClause?.namedBindings;
+  if (bindings === undefined) return null;
```

Comment:
Nit: always use braces:
```suggestion
  if (bindings === undefined) {
    return null;
  }
```